scripts/metric_reporter/junit_xml_parser.py

```diff
@@ -0,0 +1,179 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+"""Module for parsing test suite results from JUnit XML content."""
+
+import logging
+import re
+from pathlib import Path
+from typing import Any
+
+import defusedxml.ElementTree as ElementTree
+from pydantic import BaseModel, ValidationError
+
+from scripts.common.error import BaseError
+
+
+class JUnitXMLProperty(BaseModel):
+    """Represents a property of a test case."""
+
+    name: str
+    value: str
+
+
+class JUnitXMLSkipped(BaseModel):
+    """Represents a skipped test case."""
+
+    reason: str | None = None
+
+
+class JUnitXMLFailure(BaseModel):
+    """Represents a failure of a test case."""
+
+    message: str
+    type: str | None = None
+    text: str | None = None
+
+
+class JUnitXMLSystemOut(BaseModel):
+    """Represents system out information."""
+
+    text: str | None = None
+
+
+class JUnitXMLTestCase(BaseModel):
+    """Represents a test case in a test suite."""
+
+    name: str
+    classname: str | None = None
+    time: float | None = None
+    properties: list[JUnitXMLProperty] | None = None
+    skipped: JUnitXMLSkipped | None = None
+    failure: JUnitXMLFailure | None = None
+    system_out: JUnitXMLSystemOut | None = None
+
+
+class JUnitXMLTestSuite(BaseModel):
+    """Represents a test suite containing multiple test cases."""
+
+    name: str
+    timestamp: str | None = None
+    hostname: str | None = None
+    tests: int
+    failures: int
+    skipped: int | None = None
+    time: float | None = None
+    errors: int | None = None
+    test_cases: list[JUnitXMLTestCase]
+
+
+class JUnitXMLTestSuites(BaseModel):
+    """Represents a collection of test suites."""
+
+    id: str | None = None
+    name: str | None = None
+    tests: int | None = None
+    failures: int | None = None
+    skipped: int | None = None
+    errors: int | None = None
+    time: float | None = None
+    timestamp: str | None = None
+    test_suites: list[JUnitXMLTestSuite] = []
+
+
+class JUnitXMLJobTestSuites(BaseModel):
+    """Represents the test suite results for a CircleCI job."""
+
+    job: int
+    test_suites: list[JUnitXMLTestSuites]
+
+
+class JUnitXmlParserError(BaseError):
```

Comment:
Should this be `JUnitXMLParserError`? Sometimes you have XML in all caps and sometimes not.

Reply:
Nice! I went through and changed to Xml see https://github.com/mozilla/ecosystem-test-scripts/pull/6/commits/aa52b6510a0c42be333bcb511173989b452179b8